feat: Implement file methods

OS-pedrogustavobilro opened the pull request:
## Description

Implement all methods related to accessing Android file system:
- `createFile()` + `createDirectory` (aka mkdir)
- `readFile()` + `readFileInChunks()`
- `saveFile()` (write or append)
- `getFileUri()` (aka resolveUri)
- `getMetadata()` (aka stat)
- `listDirectory()` (aka readdir)
- `delete()` + `copy()` + `move()` (aka rename) - they apply to both files and directories

The `IONFileController` is exposed to the clients, and actual implementation is spread across different `Helper` classes.

The structure is similar to the [Geolocation Native library](https://github.com/ionic-team/ion-android-geolocation), a few differences being:

1. The controller is mostly a wrapper of the helpers, has no additional logic.
2. The method that returns a flow (`readFileInChunks`), does not return a `Result` encapsulated in it, as it's fairly redundant. Errors can be treated by calling the `catch` operator on the flow.

## Type of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply -->
- [ ] Fix (non-breaking change which fixes an issue)
- [X] Feature (non-breaking change which adds functionality)
- [ ] Refactor (cosmetic changes)
- [X] Breaking change (change that would cause existing functionality to not work as expected)

## Platforms affected
- [X] Android
- [ ] iOS
- [ ] JavaScript

## Tests

Around a hundred unit tests or so were written to test functionality in the library.

Some of the tests use Robolectric because they involve accessing a lot of Android-specific code, and the functional mocks the framework offers make it easier to test.
In other tests where we don't need to mock as much Android code, we don't use Robolectric, as tests should run quicker that way.

Other than that, manual tests were made with consumers of this library (Cordova and Capacitor plugins, outside this repo) on sample apps.

OS-pedrogustavobilro committed:
feat: Barebones createDirectory and createFile

May need some additional logic depending to vary implementation with the type of uri / filePath

OS-pedrogustavobilro committed:
feat: Barebones delete (file and directory)

May need some additional logic depending to vary implementation with the type of uri / filePath.

Also includes some minor refactors for create

OS-pedrogustavobilro committed:
feat: Barebones save and read file and directory

May need some additional logic for reading files with content://scheme, perhaps in a separate Helper

OS-pedrogustavobilro committed:
fix: Catch file and IO errors

Also includes some minor refactors

OS-pedrogustavobilro committed:
refactor: Separate helpers in different classes

OS-pedrogustavobilro committed:
refactor: Remove some exception duplication

OS-pedrogustavobilro committed:
refactor: File rename to maintain convention

OS-pedrogustavobilro committed:
feat: implement fileMetadata/stat and listDir

Also includes some refactor to test classes for better reusability

OS-pedrogustavobilro committed:
feat: Implementation for getFileUri / resolveFileSystemUri

Still missing tests, so some of the logic may change if any issues are found.

OS-pedrogustavobilro committed:
fix: Minor bugs on OSFLSTFUriHelper

Discovered in tests OSFLSTFUriHelperTest

OS-pedrogustavobilro committed:
fix: Add a few missing mime types for local files

Based on logic from Cordova's CordovaResourceApi#getMimeTypeFromPath

OS-pedrogustavobilro committed:
refactor: Move fullPath out of options

To better distinguish between content and file path Uri's

OS-pedrogustavobilro committed:
refactor: Re-order helper methods

OS-pedrogustavobilro committed:
refactor: File name

OS-pedrogustavobilro committed:
refactor: Reading/deleting files via android's ContentResolver

Includes unit tests with Robolectric

OS-pedrogustavobilro committed:
Merge branch 'development' into feat/implement-file-methods

# Conflicts:
#	src/main/kotlin/io/ionic/libs/osfilesystemlib/controller/OSFLSTController.kt
#	src/test/java/io.ionic.libs/ionfilesystemlib/ExampleUnitTest.kt

OS-pedrogustavobilro committed:
refactor: rename classes from OS to ION

OS-pedrogustavobilro committed:
refactor: Inject Context in IONFLSTFUriHelper constructor

OS-pedrogustavobilro committed:
refactor: Separate non-junit test logic from base class

OS-pedrogustavobilro committed:
fix: Close content:// stream after reading

OS-pedrogustavobilro committed:
feat: Implement IONFLSTController

Connecting the several helper classes

OS-pedrogustavobilro committed:
feat: Implement copy and rename files and directories

Includes support for copying content:// files

OS-pedrogustavobilro committed:
fix: Simplify save and create parameters

exclusive should always be true, and createFileRecursive can't be null

OS-pedrogustavobilro committed:
feat: implement readyFileByChunks

OS-pedrogustavobilro committed:
refactor: Rename directory

OS-pedrogustavobilro committed:
fix: Provide uri in metadata

Cordova and capacitor consumers will get this

OS-pedrogustavobilro committed:
fix: chunkSize in base64

OS-pedrogustavobilro committed:
chore: Add parsing of encoding from string

OS-pedrogustavobilro committed:
chore: Additional parsing for encoding and folder types

OS-pedrogustavobilro committed:
refactor: Rename IONFLST to IONFILE

To be consistent with iOS: https://github.com/ionic-team/ion-ios-filesystem/pull/1

OS-pedrogustavobilro committed:
refactor: Made helpers internal

`IONFILEController` cannot be instantiated with custom helpers outside this library/module.

OS-pedrogustavobilro committed:
fix: Do not emit chunk if file is empty

To align with iOS.

OS-pedrogustavobilro committed:
fix: Return null createdTimestamp if not available

To maintain existing behavior in capacitor

OS-pedrogustavobilro committed:
refactor: reorganize exceptions order

OS-pedrogustavobilro committed:
fix: Return permission flag when resolving uri

To make it easier to understand if certain uri's require storage permission or not.

OS-pedrogustavobilro committed:
chore: Removed unnecessary annotation

OS-pedrogustavobilro committed:
refactor: Rename permission variable to external storage

It is more to the point, and callers can use this information to check for permission or do other actions (such as updating file on MediaStore)

OS-pedrogustavobilro committed:
fix: Controller result in copy/move

OS-pedrogustavobilro committed:
fix: Creation error message

OS-pedrogustavobilro committed:
fix: Fallback in case rename/move fails

OS-pedrogustavobilro committed:
fix: Context preservation in flow

To avoid this error when collecting the flow:

java.lang.IllegalStateException: Flow invariant is violated:
		Flow was collected in [ProducerCoroutine{Active}@3d8c793, Dispatchers.IO],
		but emission happened in [kotlinx.coroutines.UndispatchedMarker@ecdf3d0, UndispatchedCoroutine{Active}@da5fc9, Dispatchers.IO].
		Please refer to 'flow' documentation or use 'flowOn' instead

OS-pedrogustavobilro committed:
refactor: rename readFileByChunks to readFileInChunks

OS-pedrogustavobilro committed:
docs: Update README.md and add Dokka

OS-pedrogustavobilro committed:
docs: Update mention of dokka

alexgerardojacinto commented on README.md:
```diff
@@ -1,9 +1,8 @@
-Documentation on how to use the library will be added below as the library gets developed.
+The `ion-android-filesystem` is a library built using `Kotlin` that allows management of files on Android applications.
 
+All file system methods are accessible via `IONFile`
```
Shouldn't this be `IONFILEController`?

OS-pedrogustavobilro replied:
My mistake there 😅  thanks

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/IONFILEController.kt:
```diff
@@ -0,0 +1,286 @@
+     */
+    suspend fun delete(uri: IONFILEUri, options: IONFILEDeleteOptions): Result<Unit> =
+        uriHelper.useUriIfResolved(uri) { resolvedUri ->
+            when {
```
Not super important, but this `when` could be refactored to:

```
when (resolvedUri) {
    is IONFILEUri.Resolved.Local -> {
        val path = resolvedUri.fullPath
        if (resolvedUri.type == LocalUriType.DIRECTORY) {
            directoriesHelper.deleteDirectory(path, options)
        } else {
            localFilesHelper.deleteFile(path)
        }
    }
    else -> contentResolverHelper.deleteFile(resolvedUri.uri)
}

```

OS-pedrogustavobilro replied:
This is maybe a matter of personal preference xD but I think here it's "clearer" to call each helper in a separate `when` clause rather than having an `if` nested in the `when`

alexgerardojacinto replied:
Yeah, I guess the only downside with the current approach is that we're duplicating the `resolvedUri is IONFILEUri.Resolved.Local` check, which is unnecessary, while with the proposed approach we avoid that.

But we can keep it how it is, I don't have a strong preference about this.

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILEContentHelper.kt:
```diff
@@ -0,0 +1,227 @@
+    /**
+     * Reads the contents of a file in chunks
+     *
+     * Useful when the file does not fit in entirely memory.
```
```suggestion
     * Useful when the file does not fit entirely in memory.
```

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILEContentHelper.kt:
```diff
@@ -0,0 +1,227 @@
+                onChunkRead = { chunk -> emit(chunk) }
+            )
+        } ?: throw IONFILEExceptions.UnknownError()
+    }.catch { throw it.mapError() }
```
Do we need to run .catch `{ throw it.mapError() }` inside of the IO thread? Probably not right? So we could do:

```
flow {
    // read file in chunks
}.flowOn(Dispatchers.IO)
.catch { throw it.mapError() }
```

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILEContentHelper.kt:
```diff
@@ -0,0 +1,227 @@
+     * @param uri the content uri of the file to delete
+     * @return success if file was deleted, error otherwise
+     */
+    suspend fun deleteFile(uri: Uri): Result<Unit> = withContext(Dispatchers.IO) {
```
This could be simplified to the following right?

```
suspend fun deleteFile(uri: Uri): Result<Unit> = withContext(Dispatchers.IO) {
    runCatching {
        val rowsDeleted = contentResolver.delete(uri, null, null)
        if (rowsDeleted <= 0) {
            throw IONFILEExceptions.UnknownError()
        }
    }.mapError()
}
```

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILEContentHelper.kt:
```diff
@@ -0,0 +1,227 @@
+     * @param cursor the android [Cursor] containing information about the uri
+     * @return the name of the file, or exception if not found
+     */
+    private fun getNameForContentUri(cursor: Cursor): String {
```
Simply a suggestion, and not super important, but this could maybe be refactored to the following, right?

```
private fun getNameForContentUri(cursor: Cursor): String {
    val columnNames = listOf(
        OpenableColumns.DISPLAY_NAME,
        MediaStore.MediaColumns.DISPLAY_NAME,
        DocumentsContract.Document.COLUMN_DISPLAY_NAME
    )

    val columnIndex = columnNames
        .map { cursor.getColumnIndex(it) }
        .firstOrNull { it >= 0 }

    return columnIndex?.let { cursor.getString(it) }
        ?: throw IONFILEExceptions.UnknownError()
}
```

OS-pedrogustavobilro replied:
That looks better, the only downside I see is that it's calling `getColumnIndex` for all values even though the index may have been found

I saw that kotlin has a function called [firstNotNullOfOrNull](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/first-not-null-of-or-null.html), which kind of avoids that.

```
        val columnIndex = columnNames
            .firstNotNullOfOrNull { cursor.getColumnIndex(it).takeIf { index -> index >= 0 }  }
```

What do you think?

alexgerardojacinto replied:
Yeah that looks even better!

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILEContentHelper.kt:
```diff
@@ -0,0 +1,227 @@
+     * @param cursor the android [Cursor] containing information about the uri
+     * @return the timestamp of last modification for the file, or 0 if not found
+     */
+    private fun getLastModifiedTimestampForContentUri(cursor: Cursor): Long {
```
Could be refactored like `getNameForContentUri `

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILEContentHelper.kt:
```diff
@@ -0,0 +1,227 @@
+     * @param cursor the android [Cursor] containing information about the uri
+     * @return the timestamp of creation for file, or 0 if not found
+     */
+    private fun getCreatedTimestampForContentUri(cursor: Cursor): Long {
```
Could be refactored like `getNameForContentUri`

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILEDirectoriesHelper.kt:
```diff
@@ -0,0 +1,112 @@
+     * @param fullPath full path to the directory
+     * @return success with list of metadata information for each file / sub-directory, error otherwise
+     */
+    suspend fun listDirectory(fullPath: String): Result<List<IONFILEMetadataResult>> =
```
We should also check for the case where the `path` isn't a directory right? We don't want to call `.listFiles` if the path isn't a directory.

Also, we could use mapNotNull instead for filterNotNull followed by map, right?

So it would be like this:

```
suspend fun listDirectory(fullPath: String): Result<List<IONFILEMetadataResult>> =
    withContext(Dispatchers.IO) {
        runCatching {
            val file = File(fullPath)
            if (!file.exists() || !file.isDirectory) {
                throw IONFILEExceptions.DoesNotExist()
            }
            file.listFiles()?.mapNotNull { getMetadata(it) } ?: emptyList()
        }
    }
```

OS-pedrogustavobilro replied:
> We should also check for the case where the path isn't a directory right? 

That will be done outside this repo, in the controller. I see you have another comment related to that, which I'll address shortly. But I suppose it doesn't hurt to add it here as well.

----------

As for [mapNotNull](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/map-not-null.html), I don't think it's the same as [filterNotNull](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/filter-not-null.html) + map - As the former filters the result after mapping, and the latter filters the result before mapping. Even though most likely `java.io` won't return null files, but rather just not send them in the array, it's "possible", hence this check.

alexgerardojacinto replied:
> That will be done outside this repo, in the controller. I see you have another comment related to that, which I'll address shortly. But I suppose it doesn't hurt to add it here as well.

Yes, and if we think of this library as an independent, stand-alone, library, which it should be, then we could have this check to make sure the feature is as complete as possible.

> As for [mapNotNull](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/map-not-null.html), I don't think it's the same as [filterNotNull](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin.collections/filter-not-null.html) + map - As the former filters the result after mapping, and the latter filters the result before mapping. Even though most likely java.io won't return null files, but rather just not send them in the array, it's "possible", hence this check.

Still, I believe the result will be the same right? And using `mapNotNull` would also be more efficient, as it does both filtering and mapping in a single step. However, I don't have a strong preference, and if you think we should keep the original code, I'm totally fine with that.

OS-pedrogustavobilro replied:
> Yes, and if we think of this library as an independent, stand-alone, library, which it should be, then we could have this check to make sure the feature is as complete as possible.

Yap I added it


> Still, I believe the result will be the same right?

Don't think so because basically `mapNotNull` is `map` followed `filterNotNull`, and I want to filter the list for null `File`'s initially.

alexgerardojacinto replied:
Yeah I get that, although I believe the result would still be the same, in practice.

But I'm totally fine with leaving it as is. I'll resolve the comment.

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILELocalFilesHelper.kt:
```diff
@@ -0,0 +1,228 @@
+            if (!file.exists()) {
+                throw IONFILEExceptions.DoesNotExist()
+            }
+            val inputStream = FileInputStream(file)
```
This could be refactored to the following, making the code cleaner, right?

```
suspend fun readFile(
        fullPath: String,
        options: IONFILEReadOptions
    ): Result<String> = withContext(Dispatchers.IO) {
        runCatching {
            val file = File(fullPath)
            if (!file.exists()) {
                throw IONFILEExceptions.DoesNotExist()
            }
            FileInputStream(file).use { inputStream ->
                inputStream.readFull(options)
            }
        }
    }
```

OS-pedrogustavobilro replied:
The Input stream is "used" inside `readFull` function, so I don't think it should be opened twice.

we can write it as a oneliner if you feel it helps readability

```kotlin
FileInputStream(file).readFull(options)
```

alexgerardojacinto replied:
That's right, I missed the `use` keyword in the readFull function. Yeah I think the oneliner makes it more readable.

OS-pedrogustavobilro replied:
Wait now that I see it, it is not, so yeah I'll call it like you mentioned

alexgerardojacinto replied:
Yeah, it's only in the `InputStreamReader` right?

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILELocalFilesHelper.kt:
```diff
@@ -0,0 +1,228 @@
+                )
+                createFileResult.exceptionOrNull()?.let { throw it }
+            }
+            val fileStream = FileOutputStream(file, options.mode == IONFILESaveMode.APPEND)
```
To make sure the FileOutputStream is closed correctly, we could use `use` right?

```
FileOutputStream(file, options.mode == IONFILESaveMode.APPEND).use { fileStream ->
    if (options.encoding is IONFILEEncoding.WithCharset) {
```

The same could apply to `BufferedOutputStream` in line 153. Although we're already using `use` on line 151, it could be refactored calling `use` like this: 

```
BufferedWriter(OutputStreamWriter(fileStream, options.encoding.charset)).use { writer ->
    writer.write(options.data)
}
```

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILELocalFilesHelper.kt:
```diff
@@ -0,0 +1,228 @@
+                writer.use { writer.write(options.data) }
+            } else {
+                val outputStream = BufferedOutputStream(fileStream)
+                val dataToDecode = if (options.data.contains(",")) {
```
Could this be refactored to the following?

`val dataToDecode = options.data.substringAfter(",", options.data)`

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILEUriHelper.kt:
```diff
@@ -0,0 +1,173 @@
+        return@withContext IONFILEUri.Resolved.Local(
+            fullPath = localFileObject.path,
+            uri = fileUri,
+            type = try {
```
Can we extract this block of code into a separate function to improve readability?

```
private fun getLocalUriType(localFileObject: File): LocalUriType {
    return try {
        when {
            !localFileObject.exists() -> LocalUriType.UNKNOWN
            localFileObject.isDirectory -> LocalUriType.DIRECTORY
            localFileObject.isFile -> LocalUriType.FILE
            else -> LocalUriType.UNKNOWN
        }
    } catch (secEx: SecurityException) {
        LocalUriType.UNKNOWN
    }
}
```

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/IONFILEUriHelper.kt:
```diff
@@ -0,0 +1,173 @@
+        localPath: String,
+        assumeExternalStorage: Boolean? = null
+    ): IONFILEUri.Resolved.Local = withContext(Dispatchers.IO) {
+        val localFileObject = if (parentFolderFileObject != null) {
```
Not super important but we could improve readability by extracting this to a separate function:

`private fun getLocalFileObject(parentFolderFileObject: File?, localPath: String): File {
    return parentFolderFileObject?.let { File(it, localPath) } ?: File(localPath)
}`

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/common/IONFILECommon.kt:
```diff
@@ -0,0 +1,167 @@
+ * @param isDirectory true if meant to create directory, false if meant to create file
+ * @return success if the file/directory was created successfully, error otherwise
+ */
+internal fun createDirOrFile(
```
This may be more a matter of opinion and it isn't super important but this could possibly be refactored to use the `when` clause instead of multiple `if` statements, like so:

```
val file = File(fullPath)
when {
        file.exists() -> throw IONFILEExceptions.CreateFailed.AlreadyExists()
        !checkParentDirectory(file, create = options.recursive) -> throw IONFILEExceptions.CreateFailed.NoParentDirectory()
        isDirectory && !file.mkdir() -> throw IONFILEExceptions.UnknownError()
        !isDirectory && !file.createNewFile() -> throw IONFILEExceptions.UnknownError()
    }
```

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/common/IONFILECommon.kt:
```diff
@@ -0,0 +1,167 @@
+    if (resolvedExtension == null) {
+        // consider a few extensions that may be missing from android; otherwise return fallback
+        resolvedExtension = when (extension) {
+            "3ga" -> "audio/3gpp"
```
Should these be moved into `const val` variables like `FILE_MIME_TYPE_FALLBACK`, just for consistency?

OS-pedrogustavobilro replied:
I refactored `FILE_MIME_TYPE_FALLBACK` to a constant because it's used across multiple files, but sure I can extract these as well.

alexgerardojacinto replied:
Yeah I figured that was the reason. But yeah I think it doesn't hurt to extract them all into constants.

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/common/IONFILEInputStreamExtensions.kt:
```diff
@@ -0,0 +1,90 @@
+ * @param options the options for configuring how to read from the stream
+ * @return the full contents of the stream
+ */
+internal fun InputStream.readFull(options: IONFILEReadOptions): String =
```
Not super important but could be refactored to:

```
internal fun InputStream.readFull(options: IONFILEReadOptions): String {
    return if (options.encoding is IONFILEEncoding.WithCharset) {
        InputStreamReader(this, options.encoding.charset).use { it.readText() }
    } else {
        Base64.encodeToString(this.readBytes(), Base64.NO_WRAP)
    }
}
```

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/common/IONFILEInputStreamExtensions.kt:
```diff
@@ -0,0 +1,90 @@
+        }
+    var bytesRead: Int
+    var readCount = 0
+    do {
```
This code could all be put inside of a `use` block from `InputStream` to make sure it is closed properly at the end, right?

OS-pedrogustavobilro replied:
Per discussion on https://github.com/ionic-team/ion-android-filesystem/pull/1#discussion_r1961855577, I believe it makes more sense for the caller to take care of closing the `InputStream`. I added a few comments on the methods to make that clearer

alexgerardojacinto replied:
Yeah that makes total sense.

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/common/IONFILEInputStreamExtensions.kt:
```diff
@@ -0,0 +1,90 @@
+    bufferSize: Int,
+    onChunkRead: suspend (String) -> Unit,
+) {
+    val chunkSize = minOf(options.chunkSize, available())
```
Not super important but the code to calculate `chunkSize` could me moved to a separate function to improve the readability of this one.

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/common/IONFILEInputStreamExtensions.kt:
```diff
@@ -0,0 +1,90 @@
+                Base64.encodeToString(byteArrayToConvert, Base64.NO_WRAP)
+            }
+            onChunkRead(readChunk)
+            readCount += bytesRead
```
Do we need this `readCount` variable? It isn't being used right?

OS-pedrogustavobilro replied:
Nicely spotted, there was a time where that `readCount` was being used, but not anymore xD

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/common/IONFILEInputStreamExtensions.kt:
```diff
@@ -0,0 +1,90 @@
+ *  or if the inputStream has less than bufferSize bytes left to read.
+ * @return number of bytes read; if or less bytes were read, it means that the end of file has been reached
+ */
+private fun InputStream.readChunk(byteArray: ByteArray, bufferSize: Int): Int {
```
This could be refactored to the following, making it easier to understand, right?

```
private fun InputStream.readChunk(byteArray: ByteArray, bufferSize: Int): Int {
    var totalBytesRead = 0
    while (totalBytesRead < byteArray.size) {
        val len = minOf(byteArray.size - totalBytesRead, bufferSize)
        val bytesRead = read(byteArray, totalBytesRead, len)
        if (bytesRead <= 0) break
        totalBytesRead += bytesRead
    }
    return totalBytesRead
}
```

OS-pedrogustavobilro replied:
Imo it's easier to understand having the loop clause all in one place instead of having the `break` in the middle there.

Do you feel like the `do -> while` is less readable?

alexgerardojacinto replied:
Hmm I think it's a matter of preference, but you might be right. I think we can leave it as is.

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/common/IONFILEUriResolveExtensions.kt:
```diff
@@ -0,0 +1,67 @@
+    uri: IONFILEUri,
+    onResolved: (IONFILEUri.Resolved.Local) -> Result<T>
+): Result<T> = useUriIfResolvedAsLocal(uri) { resolvedUri ->
+    if (resolvedUri.type != LocalUriType.FILE) {
```
Just to make sure I understand, why don't we do `resolvedUri.type == LocalUriType.DIRECTORY)` instead?

OS-pedrogustavobilro replied:
The `LocalUriType.UNKOWN` can be returned if `IONFILEUriHelper` cannot check if it's a file or directory.

This can happen e.g. if it's pointing to a file that doesn't exist. Rather than throwing an error in that case, it advances forward to call the method. Then, if the directory does not exist, a `DoesNotExistError` will be thrown after, but basically that responsibility is left outside this method.

You think the method needs renaming / the code comment clarified?

alexgerardojacinto replied:
Oh that makes sense. In that case I would just add a comment explaining it yeah.

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/common/IONFILEUriResolveExtensions.kt:
```diff
@@ -0,0 +1,67 @@
+    uri: IONFILEUri,
+    onResolved: (IONFILEUri.Resolved.Local) -> Result<T>
+): Result<T> = useUriIfResolvedAsLocal(uri) { resolvedUri ->
+    if (resolvedUri.type != LocalUriType.DIRECTORY) {
```
Same question as above, can't we check for `FILE` instead? Or do we accept it it is `UNKNOWN` ?

alexgerardojacinto commented on src/main/kotlin/io/ionic/libs/ionfilesystemlib/helper/common/IONFILEUriResolveExtensions.kt:
```diff
@@ -0,0 +1,67 @@
+    }
+}
+
+internal suspend inline fun <reified T> IONFILEUriHelper.useUriIfResolved(
```
We could make it simpler as follows right?

```
internal suspend inline fun <reified T> IONFILEUriHelper.useUriIfResolved(
    uri: IONFILEUri,
    onResolved: (IONFILEUri.Resolved) -> Result<T>
): Result<T> {
    val resolvedUri = when (uri) {
        is IONFILEUri.Resolved -> uri
        is IONFILEUri.Unresolved -> resolveUri(uri).getOrElse {
            return Result.failure(it ?: NullPointerException())
        }
    }
    return onResolved(resolvedUri)
}
```

OS-pedrogustavobilro committed:
docs: Update README and code documentation

Following comments on PR.

OS-pedrogustavobilro committed:
chore: increase JVM heap size.

To attempt to fix "Java heap space" error in CI

OS-pedrogustavobilro committed:
refactor: Readability improvements

Following PR comments

OS-pedrogustavobilro committed:
chore: Comment on reasoning for useUriIfResolved methods

OS-pedrogustavobilro committed:
refactor: Readability improvement

Following PR comment https://github.com/ionic-team/ion-android-filesystem/pull/1/files#r1961782799

OS-pedrogustavobilro committed:
refactor: Add more information to exceptions

References: https://outsystemsrd.atlassian.net/browse/RMET-3908
